src/systems/spacecraft_thruster_model/SpacecraftThrusterModel.cc

```diff
@@ -0,0 +1,431 @@
+/*
+ * Copyright 2015 Fadri Furrer, ASL, ETH Zurich, Switzerland
+ * Copyright 2015 Michael Burri, ASL, ETH Zurich, Switzerland
+ * Copyright 2015 Mina Kamel, ASL, ETH Zurich, Switzerland
+ * Copyright 2015 Janosch Nikolic, ASL, ETH Zurich, Switzerland
+ * Copyright 2015 Markus Achtelik, ASL, ETH Zurich, Switzerland
+ * Copyright 2016 Geoffrey Hunter <gbmhunter@gmail.com>
+ * Copyright (C) 2024 Open Source Robotics Foundation
+ * Copyright (C) 2022 Benjamin Perseghetti, Rudis Laboratories
+ * Copyright (C) 2024 Pedro Roque, DCS, KTH, Sweden
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "SpacecraftThrusterModel.hh"
+
+#include <mutex>
+#include <string>
+#include <optional>
+#include <chrono>
+
+#include <gz/msgs/actuators.pb.h>
+
+#include <gz/common/Profiler.hh>
+
+#include <gz/plugin/Register.hh>
+#include <gz/transport/Node.hh>
+
+#include <gz/math/Helpers.hh>
+#include <gz/math/Pose3.hh>
+#include <gz/math/Vector3.hh>
+#include <gz/msgs/Utility.hh>
+
+#include <sdf/sdf.hh>
+
+#include "gz/sim/components/Actuators.hh"
+#include "gz/sim/components/ExternalWorldWrenchCmd.hh"
+#include "gz/sim/components/JointAxis.hh"
+#include "gz/sim/components/JointVelocity.hh"
+#include "gz/sim/components/JointVelocityCmd.hh"
+#include "gz/sim/components/LinearVelocity.hh"
+#include "gz/sim/components/ParentLinkName.hh"
+#include "gz/sim/components/Pose.hh"
+#include "gz/sim/components/Wind.hh"
+#include "gz/sim/Link.hh"
+#include "gz/sim/Model.hh"
+#include "gz/sim/Util.hh"
+
+#define DEBUG 0
+
+using namespace gz;
+using namespace sim;
+using namespace systems;
+
+class gz::sim::systems::SpacecraftThrusterModelPrivate
+{
+  /// \brief Callback for actuator commands.
+  public: void OnActuatorMsg(const msgs::Actuators &_msg);
```

Comment:
Its worth pointing out that perhaps we should use the ECM for some of this. It was a mistake made by myself to implement the thruster using messages. If we had used the ECM a lot of our tests would be much more deterministic.

Reply:
In our case we are using these messages from PX4 GZ Bridge, so I'll have to keep it, or otherwise I'd have to re-write quite a bit of code on the other stack. We can keep it like this for now, perhaps? The module should still accept inputs from components - check the top of UpdateForcesAndMoments.